### DISCUSSION TITLE
Can wsl2 support a higher version of OpenGL?

### DISCUSSION THREAD

**Discussion by rjsgml5698** [microsoft/WSL#6154] [General] [answered]
Hello,

https://github.com/intel-isl/Open3D-ML/issues/134

I have the following problem.

Can wsl2 support a higher version of OpenGL(At least, 4.1 but current version is 1.4)?

**Comment by therealkenc** [comment, marked as answer]
From the link you are using `LIBGL_ALWAYS_INDIRECT=1`. The person (if any) who directed you to do that was proffering dubious guidance. [GLX](https://en.wikipedia.org/wiki/GLX) supports OpenGL 1.4. Later incarnations of OpenGL introduced memory mapped buffers and are not supported by the GLX wire protocol.

You can use mesa's llvmpipe which _very_ [recently](https://www.phoronix.com/scan.php?page=news_item&px=OpenGL-4.5-LLVMpipe-Lands) began to support OpenGL 4.x in mesa 20.2. Ubuntu 20.04 (focal) currently ships mesa 20.0.8, but you can install bleeding edge versions using kiask's [PPA](https://launchpad.net/~kisak/+archive/ubuntu/kisak-mesa). Or compile it yourself.

First-party support for GUI apps on WSL is #938.

![image](https://user-images.githubusercontent.com/7049700/97351652-4b50c680-184f-11eb-9aa7-14ad17546ec0.png)

**Reply by Trass3r** [reply]
I found https://launchpad.net/~oibaf/+archive/ubuntu/graphics-drivers/ to be even better.

**Reply by danielbisar** [reply]
@therealkenc Thanks for this post. Really nice. Works as expected. Great.

**Reply by Sasino97** [reply]
@Trass3r It worked like a charm, thank you

**Reply by Marqin** [reply]
@Trass3r seems your link doesn't support Ubuntu 20.04 anymore (even though it's LTS) :( and Kiask doesn't support OpenGL 4.5. Are there any alternatives? Is there some archive of Oibaf for 20.04?

**Comment by rjsgml5698** [comment]
@therealkenc Thanks for your reply.

I solved the problem! Thank you.

**Reply by the-m4a** [reply]
for those of us that are not well versed at this stuff (like myself)... what exactly did you have to do?  I can't seem to figure out how to get the newer version -- still only see 3.3 on my setup.  (thanks!)

```
name of display: 172.22.16.1:0
display: 172.22.16.1:0  screen: 0
direct rendering: Yes
Extended renderer info (GLX_MESA_query_renderer):
    Vendor: VMware, Inc. (0xffffffff)
    Device: llvmpipe (LLVM 10.0.0, 128 bits) (0xffffffff)
    Version: 20.0.8
    Accelerated: no
    Video memory: 7931MB
    Unified memory: no
    Preferred profile: core (0x1)
    Max core profile version: 3.3
    Max compat profile version: 3.1
    Max GLES1 profile version: 1.1
    Max GLES[23] profile version: 3.1
OpenGL vendor string: VMware, Inc.
OpenGL renderer string: llvmpipe (LLVM 10.0.0, 128 bits)
OpenGL core profile version string: 3.3 (Core Profile) Mesa 20.0.8
OpenGL core profile shading language version string: 3.30
OpenGL core profile context flags: (none)
OpenGL core profile profile mask: core profile

OpenGL version string: 3.1 Mesa 20.0.8
OpenGL shading language version string: 1.40
OpenGL context flags: (none)

OpenGL ES profile version string: OpenGL ES 3.1 Mesa 20.0.8
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 3.10
```

**Reply by Trass3r** [reply]
Your mesa is too old.
Try https://launchpad.net/~oibaf/+archive/ubuntu/graphics-drivers/.

**Comment by Trass3r** [comment]
I had to disable wgl in VcXsrv to make it work.

**Reply by Luke44C** [reply]
I couldn't find the WGL setting in GWSL.  Following your [initial suggestion](https://github.com/microsoft/WSL/discussions/6154#discussioncomment-128526), plus switching to VcXsrv and disabling WGL solved it for me.  Thank you!

**Reply by wangmiaowei** [reply]
How to swicht to VcXsrv? I can not get a Gnome enviroment in my Ubuntu wsl2 18.04 settting!

**Comment by the-m4a** [comment]
Hah!  Thank you, @Trass3r  - adding the repo from the link you'd pasted (https://launchpad.net/~oibaf/+archive/ubuntu/graphics-drivers/), doing the update / upgrade command upped those OpenGL to 4.5.  Excellent, thank you!

**Reply by Trass3r** [reply]
Yeah it supports groove and is even built with LTO, very nice.

**Comment by JDRanpariya** [comment]
I've added the oibaf's PPA repo and ran `sudo apt update/upgrade` but still I'm getting the following output
```name of display: 172.29.128.1:0
display: 172.29.128.1:0  screen: 0
direct rendering: No (LIBGL_ALWAYS_INDIRECT set)
OpenGL vendor string: Mesa/X.org
OpenGL renderer string: softpipe
OpenGL version string: 1.4 (3.3 (Compatibility Profile) Mesa 21.0.0-devel (git-8f342ce8b0))```

**Reply by Trass3r** [reply]
As it says `direct rendering: No (LIBGL_ALWAYS_INDIRECT set)`.
Make sure you disabled wgl in your x server and check the env vars. `LIBGL_ALWAYS_SOFTWARE=1` might also help.

**Reply by Trass3r** [reply]
I also had some other issues since Mesa 21 which I solved by `sudo apt purge mesa-vulkan-drivers`

**Reply by equisde** [reply]
hey i'm facing the same problem used https://launchpad.net/~oibaf/+archive/ubuntu/graphics-drivers/ didnt solved it 

my log...
```
[INFO   ] [Logger      ] Record log in /home/parallax/.kivy/logs/kivy_21-01-19_27.txt
[INFO   ] [Kivy        ] v2.0.0
[INFO   ] [Kivy        ] Installed at "/home/parallax/.local/lib/python3.6/site-packages/kivy/__init__.py"
[INFO   ] [Python      ] v3.6.9 (default, Oct  8 2020, 12:12:24)
[GCC 8.4.0]
[INFO   ] [Python      ] Interpreter at "/usr/bin/python3"
[INFO   ] [KivyMD      ] v0.104.1
[INFO   ] [Factory     ] 186 symbols loaded
[INFO   ] [ImageLoaderFFPy] Using ffpyplayer 4.3.2
[INFO   ] [Image       ] Providers: img_tex, img_dds, img_sdl2, img_pil, img_ffpyplayer
[INFO   ] [Window      ] Provider: sdl2
libGL error: No matching fbConfigs or visuals found
libGL error: failed to load driver: swrast
[INFO   ] [GL          ] Using the "OpenGL" graphics system
[INFO   ] [GL          ] Backend used <sdl2>
[INFO   ] [GL          ] OpenGL version <b'1.4 (4.6.0 - Build 29.20.100.9116)'>
[INFO   ] [GL          ] OpenGL vendor <b'Intel'>
[INFO   ] [GL          ] OpenGL renderer <b'Intel(R) Iris(R) Plus Graphics'>
[INFO   ] [GL          ] OpenGL parsed version: 1, 4
[CRITICAL] [GL          ] Minimum required OpenGL version (2.0) NOT found!

OpenGL version detected: 1.4

Version: b'1.4 (4.6.0 - Build 29.20.100.9116)'
Vendor: b'Intel'
Renderer: b'Intel(R) Iris(R) Plus Graphics'

Try upgrading your graphics drivers and/or your graphics hardware in case of problems.
segmentation fault.
```
anyway to solve it?

**Reply by wangmiaowei** [reply]
Didnot Solve my problem. And my settings are:
![image](https://user-images.githubusercontent.com/27915265/123944903-3032e980-d9d0-11eb-9dce-62219b213a59.png)
@therealkenc @Trass3r

**Reply by Trass3r** [reply]
Looks like you are using the recently merged D3D12 driver which *does* only support 3.3

**Reply by wangmiaowei** [reply]
I am  a newer to the WSL2. How can I solve the problem?

**Reply by Trass3r** [reply]
Maybe via `MESA_LOADER_DRIVER_OVERRIDE=llvmpipe` or `LIBGL_ALWAYS_SOFTWARE=true`.

**Reply by wangmiaowei** [reply]
Thanks for your suggestions. But when I input $ MESA_LOADER_DRIVER_OVERRIDE=llvmpipe and then glxinfo -B. I get something like Error: unable to open display 172.22.96.1:0. It really trobules me much.

**Reply by Trass3r** [reply]
Make sure it's up-to-date:
```sh
export DISPLAY=$(cat /etc/resolv.conf | grep nameserver | awk '{print $2}'):0
```

**Reply by wangmiaowei** [reply]
@Trass3r 
It seems that my problem has not solved:
name of display: 172.26.144.1:0
display: 172.26.144.1:0  screen: 0
direct rendering: Yes
Extended renderer info (GLX_MESA_query_renderer):
    Vendor: VMware, Inc. (0xffffffff)
    Device: llvmpipe (LLVM 10.0.0, 256 bits) (0xffffffff)
    Version: 20.0.8
    Accelerated: no
    Video memory: 15763MB
    Unified memory: no
    Preferred profile: core (0x1)
    Max core profile version: 3.3
    Max compat profile version: 3.1
    Max GLES1 profile version: 1.1
    Max GLES[23] profile version: 3.1
OpenGL vendor string: VMware, Inc.
OpenGL renderer string: llvmpipe (LLVM 10.0.0, 256 bits)
OpenGL core profile version string: 3.3 (Core Profile) Mesa 20.0.8
OpenGL core profile shading language version string: 3.30
OpenGL core profile context flags: (none)
OpenGL core profile profile mask: core profile

OpenGL version string: 3.1 Mesa 20.0.8
OpenGL shading language version string: 1.40
OpenGL context flags: (none)

OpenGL ES profile version string: OpenGL ES 3.1 Mesa 20.0.8
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 3.10
And I also add the link and do the update but there is no any upgraded or installed:
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
And my setting:
![1635000098(1)](https://user-images.githubusercontent.com/27915265/138561016-fb40ac39-fcbe-4587-a1e0-9a95275aeeb1.png)

**Reply by Trass3r** [reply]
Your mesa is too old probably.
Try https://launchpad.net/~oibaf/+archive/ubuntu/graphics-drivers

**Comment by tomvdon** [comment]
Does this not work with the native WSL2 gui? I've tried updating using the PPA listed here but I'm still stuck at 3.3

```
name of display: :0
NVD3D10: CPU cyclestats are disabled on client virtualization
NVD3D10: CPU cyclestats are disabled on client virtualization
display: :0  screen: 0
direct rendering: Yes
server glx vendor string: SGI
server glx version string: 1.4
server glx extensions:
client glx vendor string: Mesa Project and SGI
client glx version string: 1.4
client glx extensions:
GLX version: 1.4
GLX extensions:
Extended renderer info (GLX_MESA_query_renderer):
    Vendor: Microsoft Corporation (0xffffffff)
    Device: D3D12 (NVIDIA GeForce RTX 3070) (0xffffffff)
    Version: 21.2.0
    Accelerated: yes
    Video memory: 24377MB
    Unified memory: no
    Preferred profile: core (0x1)
    Max core profile version: 3.3
    Max compat profile version: 3.1
    Max GLES1 profile version: 1.1
    Max GLES[23] profile version: 3.0
OpenGL vendor string: Microsoft Corporation
OpenGL renderer string: D3D12 (NVIDIA GeForce RTX 3070)
OpenGL core profile version string: 3.3 (Core Profile) Mesa 21.2.0 - kisak-mesa PPA
OpenGL core profile shading language version string: 3.30
OpenGL core profile context flags: (none)
OpenGL core profile profile mask: core profile
OpenGL core profile extensions:
OpenGL version string: 3.1 Mesa 21.2.0 - kisak-mesa PPA
OpenGL shading language version string: 1.40
OpenGL context flags: (none)
OpenGL extensions:
OpenGL ES profile version string: OpenGL ES 3.0 Mesa 21.2.0 - kisak-mesa PPA
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 3.00
OpenGL ES profile extensions:
288 GLXFBConfigs:
```

**Reply by Trass3r** [reply]
The hardware accelerated d3d12 backend does not support OpenGL 4 (yet).

**Reply by LohiKarma97** [reply]
I follow Nvidia and MS recommendation to install WSL2 and WSLG. I end up in same situation. D3D12 doesn't support OpenGL 4 or later. 
So thanks to @Trass3r  :) when LIBGL_ALWAYS_SOFTWARE=true Opengl4.1 is supported.  But if my understanding is correct it means rendering is not made by HW so this most probably have a huge impact in performance.  
Anyone has update on the topic? Is there actually a driver supporting wsl2 and OpenGL4++ ? 
I currently have nvidia GPU and intel GPU. I wonder if the intel GPU could have some driver supporting openGL4 and is compatible with WSLG. 
Below is my GlxInfo when LIBGL_ALWAYS_SOFWARE=false. 
```name of display: :0
display: :0  screen: 0
direct rendering: Yes
Extended renderer info (GLX_MESA_query_renderer):
    Vendor: Microsoft Corporation (0xffffffff)
    Device: D3D12 (NVIDIA GeForce RTX 2080 Super with Max-Q Design) (0xffffffff)
    Version: 21.2.4
    Accelerated: yes
    Video memory: 24269MB
    Unified memory: no
    Preferred profile: core (0x1)
    Max core profile version: 3.3
    Max compat profile version: 3.1
    Max GLES1 profile version: 1.1
    Max GLES[23] profile version: 3.0
OpenGL vendor string: Microsoft Corporation
OpenGL renderer string: D3D12 (NVIDIA GeForce RTX 2080 Super with Max-Q Design)
OpenGL core profile version string: 3.3 (Core Profile) Mesa 21.2.4 - kisak-mesa PPA
OpenGL core profile shading language version string: 3.30
OpenGL core profile context flags: (none)
OpenGL core profile profile mask: core profile

OpenGL version string: 3.1 Mesa 21.2.4 - kisak-mesa PPA
OpenGL shading language version string: 1.40
OpenGL context flags: (none)

OpenGL ES profile version string: OpenGL ES 3.0 Mesa 21.2.4 - kisak-mesa PPA
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 3.00```

**Reply by Trass3r** [reply]
First of all I'd use a dev version of Mesa to test for any new features.
But I think nothing has changed. Perhaps Zink could be used in the future if they add Vulkan support for WSL but I don't know if that's technically possible.

**Reply by wangmiaowei** [reply]
@Trass3r @LohiKarma97  Thanks for your discussion. However, in my situation, when I typeLIBGL_ALWAYS_SOFTWARE=true , my renderer still keeps unchanged. Any suggestion:
![1635003332(1)](https://user-images.githubusercontent.com/27915265/138562708-f0f5b26c-40fb-472c-b1c0-5ffc998a5a89.png)

**Reply by Trass3r** [reply]
Yeah there are also vars to pick the backend directly but they've never worked for me.

**Comment by earlev4** [comment]
Hi. Good day! 

I am having similar problems with my NVIDIA GeForce GTX 1080 Ti. I am running Windows 11 and updated my WSL2 (Ubuntu 18.04) repository to the kisak-mesa PPA. After performing an update, my `glxinfo -B` outputs:

```
name of display: :0
display: :0  screen: 0
direct rendering: Yes
Extended renderer info (GLX_MESA_query_renderer):
    Vendor: Microsoft Corporation (0xffffffff)
    Device: D3D12 (NVIDIA GeForce GTX 1080 Ti) (0xffffffff)
    Version: 21.3.1
    Accelerated: yes
    Video memory: 43746MB
    Unified memory: no
    Preferred profile: core (0x1)
    Max core profile version: 3.3
    Max compat profile version: 3.1
    Max GLES1 profile version: 1.1
    Max GLES[23] profile version: 3.0
OpenGL vendor string: Microsoft Corporation
OpenGL renderer string: D3D12 (NVIDIA GeForce GTX 1080 Ti)
OpenGL core profile version string: 3.3 (Core Profile) Mesa 21.3.1 - kisak-mesa PPA
OpenGL core profile shading language version string: 3.30
OpenGL core profile context flags: (none)
OpenGL core profile profile mask: core profile

OpenGL version string: 3.1 Mesa 21.3.1 - kisak-mesa PPA
OpenGL shading language version string: 1.40
OpenGL context flags: (none)

OpenGL ES profile version string: OpenGL ES 3.0 Mesa 21.3.1 - kisak-mesa PPA
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 3.00
```
The output indicates the device is D3D12, which based on the previous conversation does not support OpenGL 4. I need to use [open3d](https://github.com/isl-org/Open3D), which requires OpenGL 4.4, so this is a show stopper for me using WSL2.

If anyone can offer any workarounds, I would greatly appreciate it.

Also, is there a site/page that is tracking the D3D12 developments to support OpenGL 4.

Thanks so much!

**Reply by Trass3r** [reply]
The answer is still Software Rendering.
The closest you get to a tracker is https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests?label_name%5B%5D=d3d12

**Reply by earlev4** [reply]
Thank you, @Trass3r. I appreciate the response and information.

**Reply by Trass3r** [reply]
Compute is next: https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/14342

**Comment by Trass3r** [comment]
Does anyone have performance numbers on the d3d12 driver?
Interestingly with the latest Ubuntu Impish mesa package (21.2.2) I get ~90fps for `glxgears` using the integrated Intel UHD graphics.
But if I upgrade to the newer kisak or oibaf (22.0.0-devel) packages it drops to ~10 fps. Of course it's not a good benchmark but still curious.

```
sudo add-apt-repository ppa:kisak/kisak-mesa (or ppa:oibaf/graphics-drivers)
sudo apt update && sudo apt upgrade -y
```

**Reply by danielbisar** [reply]
I yesterday ran https://www.geeks3d.com/20140304/gputest-0-7-0-opengl-benchmark-win-linux-osx-new-fp64-opengl-4-test-and-online-gpu-database/

Native Windows 11 with MSAAx8 still more than double as fast as using WSL with kisak without MSAA enabled at all.


```
name of display: :0
display: :0  screen: 0
direct rendering: Yes
Extended renderer info (GLX_MESA_query_renderer):
    Vendor: Microsoft Corporation (0xffffffff)
    Device: D3D12 (NVIDIA GeForce RTX 2060) (0xffffffff)
    Version: 21.3.4
    Accelerated: yes
    Video memory: 14120MB
    Unified memory: no
    Preferred profile: core (0x1)
    Max core profile version: 3.3
    Max compat profile version: 3.1
    Max GLES1 profile version: 1.1
    Max GLES[23] profile version: 3.0
OpenGL vendor string: Microsoft Corporation
OpenGL renderer string: D3D12 (NVIDIA GeForce RTX 2060)
OpenGL core profile version string: 3.3 (Core Profile) Mesa 21.3.4 - kisak-mesa PPA
OpenGL core profile shading language version string: 3.30
OpenGL core profile context flags: (none)
OpenGL core profile profile mask: core profile

OpenGL version string: 3.1 Mesa 21.3.4 - kisak-mesa PPA
OpenGL shading language version string: 1.40
OpenGL context flags: (none)

OpenGL ES profile version string: OpenGL ES 3.0 Mesa 21.3.4 - kisak-mesa PPA
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 3.00
```

`5.10.60.1-microsoft-standard-WSL2 #1 SMP Wed Aug 25 23:20:18 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux`

Results from WSL run
```
Module: FurMark
Score: 2490 points (FPS: 41)

Settings:
- 1920x1080 fullscreen
- antialiasing: Off
- duration: 60000 ms

Renderer:
- D3D12 (Nvidia GeForce RTX 2060)
- OpenGL: 3.1 Mesa 21.3.4 - kisak-mesa PP
```
The same on Windows native

```
---------------------------
GpuTest
---------------------------
GpuTest 0.7.0

http://www.geeks3d.com
Module: FurMark

Score: 9966 points (FPS: 166)

Settings:

- 1920x1080 fullscreen
- antialiasing: Off
- duration: 60000 ms



Renderer:
- NVIDIA GeForce RTX 2060/PCIe/SSE2
- OpenGL: 4.6.0 NVIDIA 511.23
```

**Reply by Trass3r** [reply]
How about the standard Impish package? Can you reproduce that difference?

**Reply by danielbisar** [reply]
After removing the kiask
==================================
GpuTest 0.7.0
http://www.geeks3d.com

Module: FurMark
Score: 2751 points (FPS: 45)

Settings:
- 1920x1080 fullscreen
- antialiasing: Off
- duration: 60000 ms

Renderer:
- D3D12 (NVIDIA GeForce RTX 2060)
- OpenGL: 3.1 Mesa 21.0.3
==================================

But this is 21.0.3 not 21.2.2. Will do a test with impish later during the day.

**Reply by Trass3r** [reply]
Tested again. Oibaf 22.0.0-devel performs comparably to the default Mesa 21.2.2 now, while kisak 21.3.4 performs poorly. Maybe they fixed something upstream 👍🏽

**Reply by danielbisar** [reply]
GpuTest 0.7.0
http://www.geeks3d.com

Module: FurMark
Score: 2583 points (FPS: 43)

Settings:
- 1920x1080 fullscreen
- antialiasing: Off
- duration: 60000 ms

Renderer:
- D3D12 (NVIDIA GeForce RTX 2060)
- OpenGL: 3.1 Mesa 21.2.2
==================================

**Reply by danielbisar** [reply]
So for me they are all almost the same. Though I noticed that the animation is not as smooth as with other version using 21.2.2

**Reply by danielbisar** [reply]
Tested now also oibaf on WSL2 and it gets the best results (still not at good as native but double as fast as the others):

```
==================================
GpuTest 0.7.0
http://www.geeks3d.com

Module: FurMark
Score: 4894 points (FPS: 81)

Settings:
- 1920x1080 windowed
- antialiasing: Off
- duration: 60000 ms

Renderer:
- D3D12 (NVIDIA GeForce RTX 2060)
- OpenGL: 3.3 (Core Profile) Mesa 22.0.0-devel (git-4198ca4 2022-01-22 focal-oibaf-ppa)
==================================
```

**Comment by Trass3r** [comment]
They finished GLES 3.1 features: https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/14624/diffs#1823e74a273d9960a3cb249f3201d8e4007c59d8

**Reply by Trass3r** [reply]
4.2 now: https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/14881

**Reply by desaixie** [reply]
Hi there. Does this mean on Win11 WSL2, with D3D12, we can upgrade to OpenGL ES 4.2 just with `sudo apt update && sudo apt upgrade -y`? I did so and am still at GLES 3.3. What am I missing here?

My `glxinfo -B` output:
name of display: :0
display: :0  screen: 0
direct rendering: Yes
Extended renderer info (GLX_MESA_query_renderer):
    Vendor: Microsoft Corporation (0xffffffff)
    Device: D3D12 (NVIDIA Quadro RTX 6000) (0xffffffff)
    Version: 21.2.6
    Accelerated: yes
    Video memory: 56868MB
    Unified memory: no
    Preferred profile: core (0x1)
    Max core profile version: 3.3
    Max compat profile version: 3.1
    Max GLES1 profile version: 1.1
    Max GLES[23] profile version: 3.0
OpenGL vendor string: Microsoft Corporation
OpenGL renderer string: D3D12 (NVIDIA Quadro RTX 6000)
OpenGL core profile version string: 3.3 (Core Profile) Mesa 21.2.6
OpenGL core profile shading language version string: 3.30
OpenGL core profile context flags: (none)
OpenGL core profile profile mask: core profile

OpenGL version string: 3.1 Mesa 21.2.6
OpenGL shading language version string: 1.40
OpenGL context flags: (none)

OpenGL ES profile version string: OpenGL ES 3.0 Mesa 21.2.6
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 3.00

**Reply by Trass3r** [reply]
https://launchpad.net/~oibaf/+archive/ubuntu/graphics-drivers

**Reply by Trass3r** [reply]
Though it seems like oibaf's libgl1-mesa-dri does not contain the d3d12.so after all, odd.
Edit: ah ok that's https://gitlab.freedesktop.org/mesa/mesa/-/issues/6733

**Comment by TheEndHunter** [comment]
There is another way to do this. build your own system distro(Using the wslg repo) 

look at the contributing section in WSLG repo to find out how to build.

Note: inside the wslg Dockerfile, remove all mentions of libdrm installation, while also making sure the container image is the latest version of cbl mariner's image at the time.

Note it needs the latest build of libdrm, so if you create a libdrm.wrap file in the mesa subprojects directory(after cloning steps) then point it to [here](https://gitlab.freedesktop.org/mesa/drm.git).

after that edit the libdrm section of mesa's meson.build to  include the higlighted portion
![wslgbuild](https://user-images.githubusercontent.com/8478343/174156479-77ea9865-1a9f-4897-8c3c-6936c0e2c2c7.png)

after this run the docker build command wslg asks you to and it should build with no errors, and follow the rest of the instruction on there, and you should be ready to go. 

after restarting wsl, make sure your updated to mesa 22.0+ (using https://launchpad.net/~oibaf/+archive/ubuntu/graphics-drivers if needed)

you should then see the following when you run glxinfo -B

![wslgCustom](https://user-images.githubusercontent.com/8478343/174155096-68605252-269a-4a6a-8cc9-04be036dda92.png)

NOTE: this is very likely to NOT be supported by microsoft at this time.

**Reply by Trass3r** [reply]
Why / in which cases would that be needed though?
So far it worked fine for me as long as you have the correct Windows drivers (this one bit me recently cause Windows always downgrades my drivers after I update) and a recent mesa.

**Reply by TheEndHunter** [reply]
did this more for tinkering than for actual use, just to show more than anything that if you have a bit of free time, and just want to do it or have a reason to, you can do it. at the moment i use it as the base for testing slightly more modern functions of OpenGL and shaders in a Linux environment, and having 64 bit types in GLSL without extensions is also nice.

**Comment by HeyHouBenjo** [comment]
Is this the place to be notified when OpenGL 4.3 Core (with Compute Shaders) is ready? My Host is Windows 10 with AMD Radeon RX 6600 and it supports up to OpenGL 4.6. In WSL 2 (Ubuntu 22.04) with the provided PPA by oibaf im getting 4.2 (Device: D3D12, Direct rendering: Yes)

**Reply by Trass3r** [reply]
https://www.phoronix.com/news/OpenGL-4.3-D3D12-Mesa
Mesa 24

**Reply by indevn** [reply]
This release is fantastic because it allows me to use OGL 4.6 with my RTX 4060LPT in WSL2.

**Comment by Trass3r** [comment]
Somehow I'm now getting only OpenGL 4.1 on Intel iGPU with latest driver 31.0.101.2127. Even the standard mesa version 23.2.1 on mantic ubuntu is definitely new enough for 4.2+: https://www.phoronix.com/news/OpenGL-4.2-Mesa-D3D12
I'm pretty sure I saw a higher version number in the past. So I guess it might be an issue with the intel driver.
Anyone else seeing that problem?

**Reply by indevn** [reply]
yes. when i update my mesa to verison 24, i can use ogl4.6 with my GPU 1(NVIDIA GeForce RTX 4060 Laptop GPU). but i still can only use ogl4.1 with my GPU 0(Intel(R) Iris(R) Xe Graphics).

**Reply by Trass3r** [reply]
Ok `export MESA_GL_VERSION_OVERRIDE=4.6 MESA_GLSL_VERSION_OVERRIDE=460` solved it, e.g. `glBufferStorage` works.
So it's just an incorrectly reported version for some reason.
